File: patterns/eventsourcing/csharp/EventSourcing.Domain/Entities/Booking.cs
using EventSourcing.Domain.Core;
using EventSourcing.Domain.Enums;
using EventSourcing.Domain.Events;

namespace EventSourcing.Domain.Entities;

public class Booking : IAggregate
{
    private readonly List<IDomainEvent> _events = [];

    private Booking()
    {
    }

    public Guid Id { get; private init; }
    public MeetingRoom MeetingRoom { get; private set; }
    public DateTime? StartTime { get; private set; }
    public DateTime? EndTime { get; private set; }

    public IReadOnlyList<Attendee> Attendees { get; private set; } = [];

    public IReadOnlyList<IDomainEvent> Events => _events;

    /// <summary>
    /// Creates a new Booking instance with the specified ID, start time, and end time.
    /// Meant to be called from business logic to create a new booking
    /// </summary>
    /// <returns></returns>
    public Booking(Guid id, MeetingRoom meetingRoom, DateTime startTime, DateTime endTime, IReadOnlyList<Attendee> attendees)
    {
        Id = id;
        MeetingRoom = meetingRoom;
        StartTime = startTime;
        EndTime = endTime;
        Attendees = attendees;
        _events.Add(new BookingCreated(meetingRoom, startTime, endTime, attendees));
    }

    /// <summary>
    /// Creates a Booking instance and restore the state from the provided events.
    /// Meant to be called when a booking is fetched from persistence storage
    /// </summary>
    /// <param name="id"></param>
    /// <param name="events"></param>
    /// <returns></returns>
    public Booking(Guid id, List<IDomainEvent> events)
    {
        Id = id;
        Apply(events);
    }

    public void UpdateTimeSlot(DateTime startTime, DateTime endTime)
    {
        StartTime = startTime;
        EndTime = endTime;
        _events.Add(new BookedTimeSlotChanged(startTime, endTime));
    }

    public void UpdateMeetingRoom(MeetingRoom meetingRoom)
    {
        MeetingRoom = meetingRoom;
        _events.Add(new MeetingRoomChanged(meetingRoom));
    }

    public void AddAttendee(Attendee newAttendee)
    {
        Attendees = [..Attendees, newAttendee];
        _events.Add(new AttendeeAdded(Attendees));
    }

    public void RemoveAttendee(Attendee attendee)
    {
        Attendees = [..Attendees.Where(x => x != attendee)];
        _events.Add(new AttendeeRemoved(Attendees));
    }

    private void Apply(List<IDomainEvent> events)
    {
        foreach (var @event in events)
        {
            _events.Add(@event);
            switch (@event)
            {
                case BookingCreated e: Apply(e); break;
                case BookedTimeSlotChanged e: Apply(e); break;
                case MeetingRoomChanged e: Apply(e); break;
                case AttendeeAdded e: Apply(e); break;
                case AttendeeRemoved e: Apply(e); break;
                default:
                    throw new NotImplementedException(
                        $"No Apply method found for event type '{@event.GetType().Name}'");
            }
        }
    }

    private void Apply(BookingCreated @event)
    {
        MeetingRoom = @event.MeetingRoom;
        StartTime = @event.StartTime;
        EndTime = @event.EndTime;
        Attendees = @event.Attendees;
    }

    private void Apply(BookedTimeSlotChanged @event)
    {
        StartTime = @event.StartTime;
        EndTime = @event.EndTime;
    }

    private void Apply(MeetingRoomChanged @event) =>
        MeetingRoom = @event.MeetingRoom;
    
    private void Apply(AttendeeAdded @event) =>
        Attendees = @event.Attendees;
    
    private void Apply(AttendeeRemoved @event) =>
        Attendees = @event.Attendees;
}

public record Attendee(string FirstName, string LastName, string Email);

File: patterns/eventsourcing/fsharp/EventSourcing.Domain/Entities/Booking.fs
module EventSourcing.Domain.Entities.Booking

open System

open EventSourcing.Domain.Enums.MeetingRoom

type Attendee =
    { FirstName: string
      LastName: string
      Email: string }

type Event =
    | BookingCreated of
        id: Guid *
        meetingRoom: MeetingRoom *
        startTime: DateTime *
        endTime: DateTime *
        attendees: Attendee List
    | BookedTimeSlotChanged of startTime: DateTime * endTime: DateTime
    | MeetingRoomChanged of meetingRoom: MeetingRoom
    | AttendeeAdded of attendees: Attendee list
    | AttendeeRemoved of attendees: Attendee list

type Booking =
    { Id: Guid
      MeetingRoom: MeetingRoom
      StartTime: DateTime
      EndTime: DateTime
      Attendees: Attendee list
      Events: Event list }

type Booking with
    static member Create
        (id: Guid, meetingRoom: MeetingRoom, startTime: DateTime, endTime: DateTime, attendees: Attendee list)
        =
        let initialEvent = BookingCreated(id, meetingRoom, startTime, endTime, attendees)

        { Id = id
          MeetingRoom = meetingRoom
          StartTime = startTime
          EndTime = endTime
          Attendees = attendees
          Events = [ initialEvent ] }

    static member Create(id: Guid, events: Event list) =
        match events with
        | [] -> failwithf $"Cannot create Booking with ID '%A{id}' from an empty list of events"

        | BookingCreated(id0, room, startTime, endTime, attendees) :: otherEvents ->
            if id0 <> id then
                failwithf $"Mismatched state ID during rehydration. Expected %A{id}, but first event has %A{id0}."

            let initialEvent = BookingCreated(id0, room, startTime, endTime, attendees)

            let initialState =
                { Id = id0
                  MeetingRoom = room
                  StartTime = startTime
                  EndTime = endTime
                  Attendees = attendees
                  Events = [ initialEvent ] }

            otherEvents
            |> List.fold (fun (currentState: Booking) e -> currentState.Apply(e)) initialState

        | _ ->
            failwithf
                $"Cannot create Booking with ID '%A{id}' from a list of events that does not start with BookingCreated event."


    member this.UpdateTimeSlot(newStartTime: DateTime, newEndTime: DateTime) =
        { this with
            StartTime = newStartTime
            EndTime = newEndTime
            Events =
                this.Events
                @ [ BookedTimeSlotChanged(startTime = newStartTime, endTime = newEndTime) ] }

    member this.UpdateMeetingRoom(newMeetingRoom: MeetingRoom) =
        { this with
            MeetingRoom = newMeetingRoom
            Events = this.Events @ [ MeetingRoomChanged(meetingRoom = newMeetingRoom) ] }

    member this.AddAttendee(newAttendee: Attendee) =
        let attendees = this.Attendees @ [ newAttendee ]

        { this with
            Attendees = attendees
            Events = this.Events @ [ AttendeeAdded(attendees = attendees) ] }

    member this.RemoveAttendee(attendeeToRemove: Attendee) =
        let attendees = this.Attendees |> List.filter (fun a -> a <> attendeeToRemove)

        { this with
            Attendees = attendees
            Events = this.Events @ [ AttendeeRemoved(attendees = attendees) ] }

    member private this.Apply(event: Event) =
        match event with
        | BookingCreated(_, meetingRoom0, startTime0, endTime0, attendees0) ->
            { this with
                MeetingRoom = meetingRoom0
                StartTime = startTime0
                EndTime = endTime0
                Attendees = attendees0
                Events = [ event ] }
        | BookedTimeSlotChanged(startTime0, endTime0) ->
            { this with
                StartTime = startTime0
                EndTime = endTime0
                Events = this.Events @ [ event ] }
        | MeetingRoomChanged(meetingRoom0) ->
            { this with
                MeetingRoom = meetingRoom0
                Events = this.Events @ [ event ] }
        | AttendeeAdded(attendees0) ->
            { this with
                Attendees = attendees0
                Events = this.Events @ [ event ] }
        | AttendeeRemoved(attendees0) ->
            { this with
                Attendees = attendees0
                Events = this.Events @ [ event ] }
